### PR TITLE
Update `TimeStepWizard`

### PR DESCRIPTION
**Pull request by iuryt**
Reopens #3244 and #3245 

If I understand it correctly, it does not make sense to have `min_change > 1` or `max_change < 1`. Am I right?
If so, should we add at least a warning for those cases?

Please, let me know if I 'reopened' the PR correctly.

### CODE REVIEW

**Comment by glwagner** [comment]
> If I understand it correctly, it does not make sense to have `min_change > 1` or `max_change < 1`. Am I right?
> If so, should we add at least a warning for those cases?

Right, it doesn't make sense. But I think we should `throw(ArgumentError` rather than merely warn because that can do bad things like spuriously increase your time-step.

**Comment by glwagner** [comment]
Some stuff later in the docstring is also weird. For example it says

> To use `TimeStepWizard`, adapt in a [`Callback`](@ref) and add it to a `Simulation`:

I don't know what it means to "adapt" a function in a callback. There's no adaptation, `TimeStepWizard` was _born_ for `Callback`!! Better to say that we "wrap" or "insert it" into `Callback`?

**Comment by iuryt** [comment]
> Some stuff later in the docstring is also weird. For example it says
> 
> > To use `TimeStepWizard`, adapt in a [`Callback`](@ref) and add it to a `Simulation`:
> 
> I don't know what it means to "adapt" a function in a callback. There's no adaptation, `TimeStepWizard` was _born_ for `Callback`!! Better to say that we "wrap" or "insert it" into `Callback`?

Good catch! I will work on the docstring and error after the weekend.

**Comment by iuryt** [comment]
I believe I implemented everything we discussed. Please, let me know if that's seems fine to be merged.

File: src/Simulations/time_step_wizard.jl [hunk 1]
```diff
@@ -30,20 +30,23 @@ Base.summary(wizard::TimeStepWizard) = string("TimeStepWizard(",
...
-(`min_change * old_Δt ≤ new_Δt ≤ max_change * old_Δt`).
+Callback function that adjusts the simulation time step to meet specified target values 
+for advective and diffusive Courant-Friedrichs-Lewy (CFL) numbers (`cfl` and `diffusive_cfl`), 
+subject to the limits
+
+```julia
+    max(min_Δt, min_change * previous_Δt) ≤ new_Δt ≤ min(max_Δt, max_change * previous_Δt)
+```
+
+where `new_Δt` is the new time step calculated by `TimeStepWizard`.
```
Comment (navidcy) [inline src/Simulations/time_step_wizard.jl:41]:
```suggestion
where `new_Δt` is the new time step calculated by the `TimeStepWizard`.
```

File: src/Simulations/time_step_wizard.jl [hunk 2]
```diff
@@ -30,20 +30,23 @@ Base.summary(wizard::TimeStepWizard) = string("TimeStepWizard(",
...
-satisfying `max_change` and `min_change` criteria so that the simulation's timestep
-`simulation.Δt` is not adapted "too quickly". In other words, `max_change` is the maximum
-and `min_change` is the minimum relative change of the time step and are, therefore, non-dimensional
-(`min_change * old_Δt ≤ new_Δt ≤ max_change * old_Δt`).
+Callback function that adjusts the simulation time step to meet specified target values 
+for advective and diffusive Courant-Friedrichs-Lewy (CFL) numbers (`cfl` and `diffusive_cfl`), 
+subject to the limits
+
+```julia
+    max(min_Δt, min_change * previous_Δt) ≤ new_Δt ≤ min(max_Δt, max_change * previous_Δt)
```
Comment (navidcy) [inline src/Simulations/time_step_wizard.jl:38]:
```suggestion
max(min_Δt, min_change * previous_Δt) ≤ new_Δt ≤ min(max_Δt, max_change * previous_Δt)
```

File: src/Simulations/time_step_wizard.jl [hunk 3]
```diff
@@ -67,7 +70,13 @@ function TimeStepWizard(FT=Float64;
                         min_Δt = 0.0,
                         cell_advection_timescale = cell_advection_timescale,
                         cell_diffusion_timescale = infinite_diffusion_timescale)
-
+    
+    # check if user gave max_change or min_change values that are invalid
+    if min_change >= 1.0 || max_change <= 1.0
+        throw(ArgumentError("Invalid value: min_change should be < 1 and max_change should be > 1. " *
+                            "You provided min_change = $min_change and max_change = $max_change."))
+    end
```
Comment (navidcy) [inline src/Simulations/time_step_wizard.jl:78]: is this simpler and cleaner?
```suggestion
    min_change ≥ 1 && throw(ArgumentError("min_change must be < 1"))

    max_change ≥ 1 && throw(ArgumentError("max_change must be > 1"))
```

File: src/Simulations/time_step_wizard.jl [hunk 4]
```diff
@@ -70,7 +70,13 @@ function TimeStepWizard(FT=Float64;
                         min_Δt = 0.0,
                         cell_advection_timescale = cell_advection_timescale,
                         cell_diffusion_timescale = infinite_diffusion_timescale)
-
+    
+    # check if user gave max_change or min_change values that are invalid
+    if min_change >= 1.0 || max_change <= 1.0
```
Comment (glwagner) [inline src/Simulations/time_step_wizard.jl]:
```suggestion
    if min_change >= 1 || max_change <= 1
```

Comment (glwagner) [reply src/Simulations/time_step_wizard.jl]: I think it's better to make these separate checks. Ie check `min_change` and then afterwards check `max_change`. The error message will be more specific.

Comment (navidcy) [reply src/Simulations/time_step_wizard.jl]: Yes, that aligns with what I suggested -- already implemented ;)

File: src/Simulations/time_step_wizard.jl [hunk 5]
```diff
@@ -30,20 +30,23 @@ Base.summary(wizard::TimeStepWizard) = string("TimeStepWizard(",
...
+where `new_Δt` is the new time step calculated by the `TimeStepWizard`.
 
 For more information on the CFL number, see its [wikipedia entry]
 (https://en.wikipedia.org/wiki/Courant%E2%80%93Friedrichs%E2%80%93Lewy_condition).
 
 Example
 =======
 
-To use `TimeStepWizard`, adapt in a [`Callback`](@ref) and add it to a `Simulation`:
+To use `TimeStepWizard`, insert it into a [`Callback`](@ref) and add it to a `Simulation`:
```
Comment (glwagner) [inline src/Simulations/time_step_wizard.jl:49]:
```suggestion
To use `TimeStepWizard`, insert it into a [`Callback`](@ref) and then add the `Callback` to a `Simulation`:
```

**Review by glwagner — APPROVED** [review]
Great improvement in docs and input validation

**Comment by iuryt** [comment]
One of the errors is for the Stratified Couette flow validation.
But `max_change = 1.1` in 
https://github.com/iuryt/Oceananigans.jl/blob/22e2e06b4f1451c18cfd4489a8862e28faab0759/validation/stratified_couette_flow/stratified_couette_flow.jl#L245

I suggest to add in the error output which value was given by the user.

File: src/Simulations/time_step_wizard.jl [hunk 6]
```diff
@@ -72,9 +72,9 @@ function TimeStepWizard(FT=Float64;
                         cell_diffusion_timescale = infinite_diffusion_timescale)
     
     # check if user gave max_change or min_change values that are invalid
-    min_change ≥ 1 && throw(ArgumentError("min_change must be < 1"))
-
-    max_change ≥ 1 && throw(ArgumentError("max_change must be > 1"))
+    min_change ≥ 1 && throw(ArgumentError("min_change must be < 1. You provided min_change = $min_change."))
+  
+    max_change ≥ 1 && throw(ArgumentError("max_change must be > 1. You provided max_change = $max_change."))
```
Comment (simone-silvestri) [inline src/Simulations/time_step_wizard.jl:77]:
```suggestion
    max_change ≤ 1 && throw(ArgumentError("max_change must be > 1. You provided max_change = $max_change."))
```

**Comment by iuryt** [comment]
How does that work for Julia?
For Python, if the user gives both `max_change<1` and `min_change>1`, this is going to give the `max_change` error only. Is it the same for Julia? If so, should we keep like this?

**Comment by navidcy** [comment]
> How does that work for Julia?
> For Python, if the user gives both `max_change<1` and `min_change>1`, this is going to give the `max_change` error only. Is it the same for Julia? If so, should we keep like this?

Sorry that was my bad. 

This

```
a==b && foo
```

means that foo will be executed only if a==b

**Comment by iuryt** [comment]
Ok. I was thinking that if, for some reason, the user giver both `max_change < 1` and `min_change > 1`, this is going to give a partial error. But well, the user can just try with a different `max_change` and then get the error for `min_change` and solve for both.

**Comment by glwagner** [comment]
> Ok. I was thinking that if, for some reason, the user giver both `max_change < 1` and `min_change > 1`, this is going to give a partial error. But well, the user can just try with a different `max_change` and then get the error for `min_change` and solve for both.

Note this is common. If your code has many bugs, you'll typically only be able to receive errors from one at a time.

**Review by navidcy — APPROVED** [review]

**Comment by navidcy** [comment]
@iuryt do you have write rights?

**Comment by iuryt** [comment]
> @iuryt do you have write rights?

nope

**Pull request merged**